File: src/ingest.jl
using UUIDs

"""
    ingest!(registry, source_id, path; doc_id, document_type,
            format, embed_backend, classify_backend,
            extraction_backend) -> Int

Read `path`, chunk, classify, embed (batched), and insert into the source's
DuckDB store. Returns the number of chunks actually inserted (dedup-aware).

`format` defaults to `:auto` (inferred from file extension: `.pdf` → `:pdf`,
everything else → `:markdown`). Pass `:pdf` or `:markdown` explicitly to override.
`extraction_backend` is required when format resolves to `:pdf`.
"""
function ingest!(registry::SourceRegistry, source_id::AbstractString, path::AbstractString;
                 doc_id::AbstractString,
                 document_type::Symbol,
                 format::Symbol               = :auto,
                 embed_backend::EmbeddingBackend,
                 classify_backend::ClassifyBackend,
                 extraction_backend::Union{ExtractionBackend,Nothing} = nothing)

    # ── Resolve format ────────────────────────────────────────────────────────
    if format === :auto
        ext = lowercase(splitext(String(path))[2])
        format = ext == ".pdf" ? :pdf : :markdown
    end

    # Validate extraction_backend early (before touching registry/db).
    if format === :pdf && isnothing(extraction_backend)
        error("extraction_backend is required for format=:pdf")
    end

    src = get_source(registry, source_id)

    # ── Get document text ─────────────────────────────────────────────────────
    local doc_text::String
    if format === :pdf
        pages    = extract_pages(extraction_backend, String(path))
        # Inject page markers AFTER the first heading line so they land in the
        # section body (not the preamble, which the markdown parser discards).
        # For pages whose text has no heading, the marker is prepended instead
        # and will appear in the preamble of the NEXT section's flush buffer.
        function _inject_marker(page_text::String, page_num::Int)
            marker = "<!-- page $page_num -->"
            lines  = split(page_text, '\n'; limit=2)
            if length(lines) >= 1 && startswith(strip(lines[1]), "#")
                # Insert marker on the line immediately after the heading
                rest = length(lines) == 2 ? lines[2] : ""
                return "$(lines[1])\n$(marker)\n$(rest)"
            else
                return "$(marker)\n$(page_text)"
            end
        end
        doc_text = join(
            [_inject_marker(p.text, p.page_num) for p in pages],
            "\n\n",
        )
    elseif format === :markdown
        doc_text = read(path, String)
    else
        error("unsupported format: $format (expected :pdf or :markdown)")
    end

    # ── Chunk ─────────────────────────────────────────────────────────────────
    raws = chunk_document(doc_text, src.chunking)
    isempty(raws) && return 0

    # For PDF: extract page numbers from markers, strip markers from text.
    if format === :pdf
        raws = _assign_pages(raws)
    end

    # ── Classify ──────────────────────────────────────────────────────────────
    classified = [classify(classify_backend; text=r.text, heading_path=r.heading_path)
                  for r in raws]

    # ── Embed (batched) ───────────────────────────────────────────────────────
    embeddings = embed(embed_backend, [r.text for r in raws])

    # ── Build Chunk objects ───────────────────────────────────────────────────
    chunks = Chunk[]
    for (i, r) in enumerate(raws)
        cls = classified[i]
        push!(chunks, Chunk(
            id              = string(uuid4()),
            source_id       = src.id,
            doc_id          = String(doc_id),
            doc_path        = abspath(path),
            text            = r.text,
            embedding       = embeddings[i],
            embedding_model = src.embedding_model,
            token_count     = token_count(r.text),
            content_hash    = content_hash(r.text),
            document_type   = document_type,
            system          = src.system,
            edition         = "",
            page            = r.page,
            heading_path    = r.heading_path,
            chunk_order     = r.chunk_order,
            parent_id       = nothing,
            content_type    = cls.content_type,
            tags            = cls.tags,
            move_trigger    = cls.move_trigger,
            scene_type      = cls.scene_type,
            encounter_key   = cls.encounter_key,
            npc_name        = cls.npc_name,
            license         = src.license,
        ))
    end
    return insert_chunks(src, chunks)
end

"""
    _assign_pages(chunks) -> Vector{RawChunk}

Scans each chunk's text for `<!-- page N -->` markers injected during PDF extraction.
Sets `chunk.page` to the last marker found in that chunk's text (the page where the
body content lives). Strips all markers from the returned chunk text.
"""
function _assign_pages(chunks::Vector{RawChunk})
    map(chunks) do rc
        matches = collect(eachmatch(r"<!-- page (\d+) -->", rc.text))
        page    = isempty(matches) ? "" : String(matches[end][1])
        clean   = strip(replace(rc.text, r"<!-- page \d+ -->\n?" => ""))
        RawChunk(
            heading_path = rc.heading_path,
            text         = clean,
            chunk_order  = rc.chunk_order,
            page         = page,
        )
    end
end

_backend_model_name(::MockEmbeddingBackend) = "mock"
_backend_model_name(b::OllamaBackend)       = b.model
_backend_model_name(::EmbeddingBackend)     = ""

File: src/ingest.ts
import { randomUUID } from "node:crypto";
import { resolve, extname } from "node:path";
import type {
  Chunk, DocumentType, RawChunk, Source,
} from "./types.ts";
import { chunkDocument } from "./chunker.ts";
import { contentHash, tokenCount } from "./tokenize.ts";
import type { EmbeddingBackend } from "./backends/embedding.ts";
import type { ClassifyBackend } from "./backends/classify.ts";
import type { ExtractionBackend } from "./backends/extraction.ts";
import { insertChunks } from "./storage.ts";

export interface IngestOptions {
  source: Source;
  path: string;
  docId: string;
  documentType: DocumentType;
  format?: "auto" | "markdown" | "pdf";
  embedBackend: EmbeddingBackend;
  classifyBackend: ClassifyBackend;
  extractionBackend?: ExtractionBackend;
}

export function injectPageMarker(pageText: string, pageNum: number): string {
  const marker = `<!-- page ${pageNum} -->`;
  const nl = pageText.indexOf("\n");
  const line1 = nl < 0 ? pageText : pageText.slice(0, nl);
  const rest = nl < 0 ? "" : pageText.slice(nl + 1);
  if (line1.trim().startsWith("#")) {
    return `${line1}\n${marker}\n${rest}`;
  }
  return `${marker}\n${pageText}`;
}

export function assignPages(chunks: RawChunk[]): RawChunk[] {
  return chunks.map((rc) => {
    const matches = [...rc.text.matchAll(/<!-- page (\d+) -->/g)];
    const page = matches.length === 0 ? "" : matches[matches.length - 1]![1]!;
    const clean = rc.text.replace(/<!-- page \d+ -->\n?/g, "").trim();
    return { headingPath: rc.headingPath, text: clean, chunkOrder: rc.chunkOrder, page };
  });
}

function searchText(r: RawChunk): string {
  if (r.headingPath.length === 0) return r.text;
  return `${r.headingPath.join(" > ")}\n\n${r.text}`;
}

export async function ingest(opts: IngestOptions): Promise<number> {
  let format = opts.format ?? "auto";
  if (format === "auto") {
    format = extname(opts.path).toLowerCase() === ".pdf" ? "pdf" : "markdown";
  }
  if (format === "pdf" && !opts.extractionBackend) {
    throw new Error("extractionBackend is required for format=pdf");
  }

  let docText: string;
  if (format === "pdf") {
    const pages = await opts.extractionBackend!.extractPages(opts.path);
    docText = pages.map((p) => injectPageMarker(p.text, p.pageNum)).join("\n\n");
  } else {
    docText = await Bun.file(opts.path).text();
  }

  let raws = chunkDocument(docText, opts.source.chunking);
  if (raws.length === 0) return 0;
  if (format === "pdf") raws = assignPages(raws);

  const classified = await opts.classifyBackend.classifyBatch(raws);
  const searchTexts = raws.map(searchText);
  const embeddings = await opts.embedBackend.embed(searchTexts);
  const absPath = resolve(opts.path);

  const chunks: Chunk[] = raws.map((r, i) => {
    const cls = classified[i]!;
    return {
      id: randomUUID(),
      sourceId: opts.source.id,
      docId: opts.docId,
      docPath: absPath,
      text: searchTexts[i]!,
      embedding: embeddings[i]!,
      embeddingModel: opts.source.embeddingModel,
      tokenCount: tokenCount(r.text),
      contentHash: contentHash(r.text),
      documentType: opts.documentType,
      system: opts.source.system,
      edition: "",
      page: r.page,
      headingPath: r.headingPath,
      chunkOrder: r.chunkOrder,
      parentId: null,
      contentType: cls.contentType,
      tags: cls.tags,
      moveTrigger: cls.moveTrigger,
      sceneType: cls.sceneType,
      encounterKey: cls.encounterKey,
      npcName: cls.npcName,
      license: opts.source.license,
    };
  });

  return insertChunks(opts.source, chunks);
}
